src/promql/src/planner.rs

```diff
@@ -328,53 +349,137 @@ impl PromPlanner {
+                            lhs_plan,
+                            rhs_plan,
+                            is_comparison_op,
+                            should_return_bool,
+                            *op,
+                        )?
+                    }
+                    // e.g. scalar(...) + metircs
+                    (Some(LiteralExpr::Runtime(lhs_plan)), None) => {
+                        let rhs_plan = self.prom_expr_to_plan(*rhs.clone()).await?;
+                        let left_table_ref = lhs_plan
+                            .schema()
+                            .field(0)
+                            .qualifier()
+                            .map(|x| x.to_owned_reference())
+                            .unwrap_or_else(|| OwnedTableReference::bare(""));
+                        let right_table_ref = self
+                            .table_ref()
+                            .unwrap_or_else(|_| OwnedTableReference::bare(""));
+                        // clear the tag value on lhs_plan to avoid join on tag column
+                        self.ctx.tag_columns.clear();
+                        let left_field_columns = vec![lhs_plan.schema().field(1).name().clone()];
+                        let right_field_columns = self.ctx.field_columns.clone();
+                        self.join_time_series(
+                            left_table_ref,
+                            right_table_ref,
+                            left_field_columns.into_iter(),
+                            right_field_columns.into_iter(),
+                            lhs_plan,
+                            rhs_plan,
+                            is_comparison_op,
+                            should_return_bool,
+                            *op,
+                        )?
+                    }
+                    // e.g. scalar(...) + 1
+                    (Some(LiteralExpr::Runtime(lhs_plan)), Some(LiteralExpr::Const(rhs))) => {
+                        let exprs = lhs_plan.expressions();
+                        let time_index = &exprs[0];
+                        let lhs = &exprs[1];
+                        let binary_expr_builder = Self::prom_token_to_binary_expr_builder(*op)?;
+                        let binary_expr = binary_expr_builder(lhs.clone(), rhs)?;
+                        self.ctx.field_columns = vec![binary_expr
+                            .display_name()
+                            .context(DataFusionPlanningSnafu)?];
+                        LogicalPlanBuilder::from(lhs_plan)
+                            .project(vec![time_index.clone(), binary_expr])
+                            .context(DataFusionPlanningSnafu)?
+                            .build()
+                            .context(DataFusionPlanningSnafu)?
+                    }
+                    // e.g. 1 + scalar(...)
+                    (Some(LiteralExpr::Const(lhs)), Some(LiteralExpr::Runtime(rhs_plan))) => {
+                        let exprs = rhs_plan.expressions();
```

Comment:
Considering addition conforms to the commutative property, can we reuse the above logic? Make it to be a function, and call it here with swapped arguments.

Reply:
The associativity law does not hold for subtraction, but we can indeed abstract functions, thereby reducing some repetitive code